ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -0,0 +1,414 @@
+        (validateEpoch validateEnv shouldBeFinalised epoch (Just prevHash)) >>= \case
+          Left  ()              -> cleanup lastValid epoch $> lastValid
+          Right Nothing         -> continueOrStop lastValid             epoch prevHash
+          Right (Just validBlk) -> continueOrStop (epoch, Tip validBlk) epoch prevHash'
+            where
+              prevHash' = At (snd validBlk)
+
+    -- | Validate the next epoch, unless the epoch just validated is the last
+    -- epoch to validate. Cleanup files corresponding to epochs after the
+    -- epoch in which we found the last valid block. Return that epoch and the
+    -- tip corresponding to that block.
+    continueOrStop
+      :: (EpochNo, ImmTipWithHash hash)
+      -> EpochNo         -- ^ The epoch just validated
+      -> WithOrigin hash -- ^ The hash of the last block of the previous epoch
+      -> m (EpochNo, ImmTipWithHash hash)
+    continueOrStop lastValid epoch prevHash
+      | epoch < lastEpoch
+      = go lastValid (epoch + 1) prevHash
+      | otherwise
+      = assert (epoch == lastEpoch) $ do
+        cleanup lastValid epoch
+        return lastValid
+
+    -- | Remove left over files from epochs newer than the last epoch
+    -- containing a valid file. Also unfinalise it if necessary.
+    cleanup
+      :: (EpochNo, ImmTipWithHash hash)  -- ^ The last valid epoch and tip
+      -> EpochNo  -- ^ The last validated epoch, could have been invalid or
+                  -- empty
+      -> m ()
+    cleanup (lastValidEpoch, tip) lastValidatedEpoch = case tip of
+      TipGen ->
+        removeFilesStartingFrom hasFS 0
+      Tip _  -> do
+        removeFilesStartingFrom hasFS (lastValidEpoch + 1)
+        when (lastValidEpoch < lastValidatedEpoch) $
+          Primary.unfinalise hasFS err lastValidEpoch
+
+-- | Validate the given most recent epoch. If that epoch contains no valid
+-- block, try the epoch before it, and so on. Stop as soon as an epoch with a
+-- valid block is found, returning that epoch and the tip corresponding to
+-- that block. If no valid blocks are found, epoch 0 and 'TipGen' is returned.
+validateMostRecentEpoch
+  :: forall m hash h e. (IOLike m, Eq hash, HasCallStack)
+  => ValidateEnv m hash h e
+  -> EpochNo
+     -- ^ Most recent epoch on disk, the epoch to validate
+  -> m (EpochNo, ImmTipWithHash hash)
+validateMostRecentEpoch validateEnv@ValidateEnv { hasFS } = go
+  where
+    go :: EpochNo -> m (EpochNo, ImmTipWithHash hash)
+    go epoch = runExceptT
+      (validateEpoch validateEnv ShouldNotBeFinalised epoch Nothing) >>= \case
+        Right (Just validBlk) -> do
+            -- Found a valid block, we can stop now.
+            removeFilesStartingFrom hasFS (epoch + 1)
+            return (epoch, Tip validBlk)
+        _  -- Epoch should be truncated or is empty
+          | epoch > 0 -> go (epoch - 1)
+          | otherwise -> do
+            -- Found no valid blocks on disk
+            removeFilesStartingFrom hasFS 0
```

Comment:
We need this because it might be empty, in which case we should remove its index files.

Add a comment that we could more precise in what should happen.